google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java

```diff
@@ -0,0 +1,347 @@
+  private JLabel errorIcon;
+  private JLabel errorMessage;
+  private AppEngineDeploymentConfiguration deploymentConfiguration;
+  private AppEngineHelper appEngineHelper;
+
+  FlexibleFacetEditor(@Nullable AppEngineDeploymentConfiguration deploymentConfiguration,
+      @Nullable Project project) {
+    this.appEngineHelper = new CloudSdkAppEngineHelper(project);
+    this.deploymentConfiguration = deploymentConfiguration;
+
+    yaml.addBrowseFolderListener(
+        GctBundle.message("appengine.flex.config.browse.app.yaml"),
+        null /* description */,
+        project,
+        FileChooserDescriptorFactory.createSingleFileDescriptor().withFileFilter(
+            virtualFile -> Comparing.equal(virtualFile.getExtension(), "yaml")
+                || Comparing.equal(virtualFile.getExtension(), "yml")
+        )
+    );
+
+    yaml.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent event) {
+        toggleDockerfileSection();
+        validateConfiguration();
+      }
+    });
+
+    dockerfile.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent event) {
+        toggleDockerfileSection();
+        validateConfiguration();
+      }
+    });
+
+    dockerfile.addBrowseFolderListener(
+        GctBundle.message("appengine.flex.config.browse.dockerfile"),
+        null /* description */,
+        project,
+        FileChooserDescriptorFactory.createSingleFileDescriptor()
+    );
+
+    genYamlButton.addActionListener(new GenerateConfigActionListener(project, "app.yaml",
+        appEngineHelper::defaultAppYaml, yaml, this::validateConfiguration));
+
+    genDockerfileButton.addActionListener(new GenerateConfigActionListener(project, "Dockerfile",
+        () -> appEngineHelper.defaultDockerfile(AppEngineFlexibleDeploymentArtifactType.WAR),
+        dockerfile, this::validateConfiguration
+    ));
+
+    yaml.setText(deploymentConfiguration.getYamlPath());
+    dockerfile.setText(deploymentConfiguration.getDockerFilePath());
+
+    errorIcon.setIcon(Ide.Error);
+    errorIcon.setVisible(false);
+    errorMessage.setVisible(false);
+
+    validateConfiguration();
+    toggleDockerfileSection();
+  }
+
+  @NotNull
+  @Override
+  public JComponent createComponent() {
+    return mainPanel;
+  }
+
+  @Override
+  public boolean isModified() {
+    return !yaml.getText().equals(deploymentConfiguration.getYamlPath())
+        || !dockerfile.getText().equals(deploymentConfiguration.getDockerFilePath());
+  }
+
+  @Override
+  public void reset() {
+    yaml.setText(deploymentConfiguration.getYamlPath());
+    dockerfile.setText(deploymentConfiguration.getDockerFilePath());
+
+    toggleDockerfileSection();
+  }
+
+  @Override
+  public void apply() throws ConfigurationException {
+    validateConfiguration();
+    if (!isValidConfigurationFile(yaml.getText())) {
```

Comment:
In the context of this class, validateConfiguration() doesn't throw any exceptions. It just enables/disables error messages.